ydb/core/tx/datashard/sample_k.cpp

```diff
@@ -0,0 +1,331 @@
+    };
+
+    TReallyFastRng32 Rng;
+    ui64 MaxProbability = 0;
+    std::vector<Probability> MaxRows;
+    std::vector<TString> DataRows;
+
+public:
+    static constexpr NKikimrServices::TActivity::EType ActorActivityType() {
+        return NKikimrServices::TActivity::BUILD_INDEX_SCAN_ACTOR; // TODO(mbkkt) Should be different activity type?
+    }
+
+    TSampleKScan(TAutoPtr<TEvDataShard::TEvSampleKResponse>&& response,
+                 const TActorId& responseActorId,
+                 const TSerializedTableRange& range,
+                 ui64 k,
+                 ui64 seed,
+                 TProtoColumnsCRef columns,
+                 const TUserTable& tableInfo)
+        : TActor(&TThis::StateWork)
+        , Response(std::move(response))
+        , ResponseActorId(responseActorId)
+        , ScanTags(BuildTags(tableInfo, columns))
+        , KeyTypes(tableInfo.KeyColumnTypes)
+        , TableRange(tableInfo.Range)
+        , RequestedRange(range)
+        , K(k)
+        , Rng(seed)
+    {
+    }
+
+    ~TSampleKScan() override = default;
+
+    TInitialState Prepare(IDriver* driver, TIntrusiveConstPtr<TScheme>) noexcept override {
+        TActivationContext::AsActorContext().RegisterWithSameMailbox(this);
+        auto ctx = TActivationContext::AsActorContext().MakeFor(SelfId());
+
+        LOG_DEBUG_S(ctx, NKikimrServices::TX_DATASHARD,
+                    "Prepared " << Debug());
+
+        Driver = driver;
+
+        return {EScan::Feed, {}};
+    }
+
+    EScan Seek(TLead& lead, ui64 seq) noexcept override {
+        auto ctx = TActivationContext::AsActorContext().MakeFor(SelfId());
+        LOG_TRACE_S(ctx, NKikimrServices::TX_DATASHARD,
+                    "Seek no " << seq << " " << Debug());
+        if (seq) {
+            SendResponse();
+            return EScan::Final;
+        }
+
+        auto scanRange = Intersect(KeyTypes, RequestedRange.ToTableRange(), TableRange.ToTableRange());
+
+        if (bool(scanRange.From)) {
+            auto seek = scanRange.InclusiveFrom ? NTable::ESeek::Lower : NTable::ESeek::Upper;
+            lead.To(ScanTags, scanRange.From, seek);
+        } else {
+            lead.To(ScanTags, {}, NTable::ESeek::Lower);
+        }
+
+        if (bool(scanRange.To)) {
+            lead.Until(scanRange.To, scanRange.InclusiveTo);
+        }
+
+        return EScan::Feed;
+    }
+
+    EScan Feed(TArrayRef<const TCell> key, const TRow& row) noexcept final {
+        auto ctx = TActivationContext::AsActorContext().MakeFor(SelfId());
+        LOG_TRACE_S(ctx, NKikimrServices::TX_DATASHARD,
+                    "Feed key " << DebugPrintPoint(KeyTypes, key, *AppData()->TypeRegistry)
+                                << " " << Debug());
+
+        const auto probability = Rng.GenRand64();
+        if (DataRows.size() < K) {
+            MaxRows.push_back({probability, DataRows.size()});
+            DataRows.emplace_back(TSerializedCellVec::Serialize(*row));
```

Comment:
How many columns are serialized here? All in table or only requested?

Reply:
only requested